components/Init/src/virtio_con.c

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2022, UNSW (ABN 57 195 873 179)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+#include <autoconf.h>
+
+#include <utils/util.h>
+
+#include <camkes.h>
+#include <camkes/dataport.h>
+
+#include <sel4vmmplatsupport/drivers/virtio_con.h>
+#include <sel4vmmplatsupport/device.h>
+
+#include <sel4vm/guest_irq_controller.h>
+#include <sel4vm/boot.h>
+
+#include <platsupport/serial.h>
+
+#include <virtqueue.h>
+#include <camkes/virtqueue.h>
+
+#include "virtio_con.h"
+#include "virtio_irq.h"
+
+static virtio_con_t *virtio_con;
+
+typedef struct virtio_con_cookie {
+    virtio_con_t *virtio_con;
+    vm_t *vm;
+} virtio_con_cookie_t;
+
+/* Configured in camkes */
+#define NUM_PORTS ARRAY_SIZE(serial_layout)
+
+/* 4088 because the serial_shmem_t struct has to be 0x1000 bytes big */
+#define BUFSIZE 4088
+
+/* This struct occupies exactly 1 page and represents the data in the shmem region */
+typedef struct serial_shmem {
+    uint32_t head;
+    uint32_t tail;
+    char buf[BUFSIZE];
+} serial_shmem_t;
+compile_time_assert(serial_shmem_1k_size, sizeof(serial_shmem_t) == 0x1000);
+
+/*
+ * Represents the virtqueues used in a given link between
+ * two VMs.
+ */
+typedef struct serial_conn {
+    void (*notify)(void);
+    volatile serial_shmem_t *recv_buf;
+    volatile serial_shmem_t *send_buf;
+} serial_conn_t;
+static serial_conn_t connections[NUM_PORTS];
+
+/* This is a temporary buffer used to stage data before completing the virtio RX */
+char temp_rx_buf[BUFSIZE];
```

Comment:
declare `static` if it's not visible outside this file; name something less likely to collide if it is meant to be visible.